File: apps/fsharp/src/MachineLearning/Chapter12/Boston.fs
module MachineLearning.Chapter12.Boston

open FSharp.Data
open MachineLearning.Chapter02.IrisPreprocessing
open MachineLearning.Chapter07.Matrix

/// 型プロバイダが列の名前と型を知るためのサンプル（架空の値）
[<Literal>]
let BostonSample =
    "CRIME,ZN,INDUS,CHAS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,PRICE\nlow,0.5,0.5,0,0.5,0.5,0.5,0.5,1,1,0.5,0.5,0.5,0.5"

/// この章で使う列は小数として読む。サンプルには空欄が無いが、使わない列には空欄があるので、
/// AssumeMissingValues で「どの列にも空欄がありうる」ものとして型を作らせる
type BostonCsv =
    CsvProvider<BostonSample, Schema="RM=float,PTRATIO=float,LSTAT=float,PRICE=float", AssumeMissingValues=true>

let FeatureNames = [ "RM"; "PTRATIO"; "LSTAT" ]

[<Literal>]
let Target = "PRICE"

/// z スコアの絶対値がこの値を超える値を持つ行を外れ値とする
[<Literal>]
let OutlierThreshold = 3.0

/// Boston.csv のうち、この章で使う列だけを列名から値への Map にする
let loadBoston (csvFile: string) : Map<string, float> list =
    BostonCsv.Load(csvFile).Rows
    |> Seq.map (fun row -> Map.ofList [ "RM", row.RM; "PTRATIO", row.PTRATIO; "LSTAT", row.LSTAT; Target, row.PRICE ])
    |> Seq.toList

/// 列の平均値と標準偏差。標準偏差は偏差の二乗和を「件数 - ddof」で割って求める
let private meanAndStd (ddof: int) (values: float list) : float * float =
    let mean = List.average values

    let squares = values |> List.sumBy (fun value -> (value - mean) * (value - mean))

    mean, sqrt (squares / float (values.Length - ddof))

/// 列ごとの z スコア（標本標準偏差で割る）の絶対値が threshold を超える値を、1 つでも持つ行を除く
let removeOutliers (columns: string list) (threshold: float) (rows: Map<string, float> list) : Map<string, float> list =
    let stats =
        columns
        |> List.map (fun column -> column, rows |> List.map (fun row -> row[column]) |> meanAndStd 1)

    let isOutlier (row: Map<string, float>) =
        stats
        |> List.exists (fun (column, (mean, std)) -> abs ((row[column] - mean) / std) > threshold)

    rows |> List.filter (isOutlier >> not)

/// 訓練データから求めた標準化の平均値・標準偏差と、2 次の項を作る列の組
type PolynomialScaler =
    {
        Columns: string list
        Stats: (float * float) list
        /// 2 次の項を作る列の番号の組（i <= j）
        Pairs: (int * int) list
        FeatureNames: string list
    }

/// 平均値と、件数で割る標準偏差（母標準偏差）を訓練データから求め、2 次の項の組を決める
let fitPolynomialScaler (columns: string list) (x: Map<string, float> list) : PolynomialScaler =
    let n = columns.Length

    let pairs =
        [
            for i in 0 .. n - 1 do
                for j in i .. n - 1 -> i, j
        ]

    let pairName (i, j) =
        if i = j then
            $"{columns[i]}^2"
        else
            $"{columns[i]} {columns[j]}"

    {
        Columns = columns
        Stats =
            columns
            |> List.map (fun column -> x |> List.map (fun row -> row[column]) |> meanAndStd 0)
        Pairs = pairs
        FeatureNames = columns @ (pairs |> List.map pairName)
    }

/// 標準化した値と、その 2 次の項を並べた行列にする
let transformPolynomial (scaler: PolynomialScaler) (x: Map<string, float> list) : Matrix =
    x
    |> List.map (fun row ->
        let z =
            List.map2 (fun column (mean, std) -> (row[column] - mean) / std) scaler.Columns scaler.Stats
            |> List.toArray

        Array.append z (scaler.Pairs |> List.map (fun (i, j) -> z[i] * z[j]) |> List.toArray))
    |> List.toArray

/// 訓練データ・検証データ・テストデータと、特徴量の名前
type BostonDataset =
    {
        XTrain: Matrix
        TTrain: float list
        XValid: Matrix
        TValid: float list
        XTest: Matrix
        TTest: float list
        FeatureNames: string list
    }

/// 外れ値を除き、テストデータを分けてから、残りを訓練データと検証データに分ける。
/// 標準化と 2 次の項は、訓練データの平均値と標準偏差で 3 つすべてに適用する
let prepareBoston (csvFile: string) (testSize: float) (validationSize: float) (seed: int) : BostonDataset =
    let rows =
        loadBoston csvFile
        |> removeOutliers (FeatureNames @ [ Target ]) OutlierThreshold

    let x = rows |> List.map (Map.remove Target)
    let t = rows |> List.map (fun row -> row[Target])
    let outer = splitTrainTest testSize seed x t
    let inner = splitTrainTest validationSize seed outer.XTrain outer.TTrain
    let scaler = fitPolynomialScaler FeatureNames inner.XTrain

    {
        XTrain = transformPolynomial scaler inner.XTrain
        TTrain = inner.TTrain
        XValid = transformPolynomial scaler inner.XTest
        TValid = inner.TTest
        XTest = transformPolynomial scaler outer.XTest
        TTest = outer.TTest
        FeatureNames = scaler.FeatureNames
    }

File: apps/ruby/lib/getting_started_ml/chapter12/boston.rb
# frozen_string_literal: true

module GettingStartedMl
  # 第 12 章の前処理。ボストンの住宅価格（Boston.csv）を、外れ値を除いて訓練・検証・テストの 3 つに分ける。
  module Chapter12
    # 特徴量の列。
    FEATURES = %w[RM PTRATIO LSTAT].freeze
    # 正解の列。
    TARGET = "PRICE"
    # z スコアの絶対値がこの値を超える行を外れ値とする。
    OUTLIER_THRESHOLD = 3.0

    # 標準化と多項式特徴量をつなげた変換。訓練データで fit し、同じ平均と標準偏差で
    # 訓練・検証・テストの 3 つを transform する。第 9 章の Standardizer と expand をそのまま使う。
    PolynomialScaler = Data.define(:standardizer) do
      # 訓練データの平均と標準偏差を覚える。
      def self.fit(x)
        new(standardizer: Chapter09::Standardizer.fit(x))
      end

      # 変換後の列名。元の列、2 乗の列（"RM^2"）、積の列（"RM LSTAT"）の順。
      def feature_names
        standardizer.columns + Chapter09.pairs_with_replacement(standardizer.columns).map(&:name)
      end

      # 標準化してから 2 次の項を足し、値だけの配列にする。
      def transform(x)
        Chapter09.expand(standardizer.transform(x), standardizer.columns).map(&:values)
      end
    end

    # 訓練・検証・テストの特徴量（値だけの配列）と正解。kept と removed は外れ値を除いたあとの件数と除いた件数。
    BostonSplit = Data.define(:x_train, :t_train, :x_valid, :t_valid, :x_test, :t_test, :feature_names, :kept, :removed)

    module_function

    # 欠損値を許さずに数値の列を読む。
    def required_number(row, column)
      row.number(column) || raise(ArgumentError, "値が空欄です: #{column}")
    end

    # 列ごとの z スコア（標準偏差は件数 n - 1 で割る標本標準偏差）の絶対値が threshold を超える値を
    # 1 つでも持つ行を除く。元の表は変えない。
    def remove_outliers(table, columns, threshold)
      stats = columns.to_h do |column|
        [column, mean_and_sample_std(table.rows.map { |row| required_number(row, column) })]
      end

      table.with(rows: table.rows.reject do |row|
        stats.any? { |column, (mean, std)| ((required_number(row, column) - mean) / std).abs > threshold }
      end)
    end

    # 平均と、件数 n - 1 で割る標準偏差の組。
    def mean_and_sample_std(values)
      mean = values.sum / values.size

      [mean, Math.sqrt(values.sum { |value| (value - mean)**2 } / (values.size - 1))]
    end

    # 外れ値を除き、全体を訓練用とテスト用に、訓練用をさらに訓練データと検証データに分ける。
    # 標準化の平均と標準偏差は訓練データだけから求め、3 つとも同じ値で変換する。
    def prepare(csv_file, test_size:, validation_size:, seed:)
      table = Chapter02::Table.load(csv_file)
      kept = remove_outliers(table, [*FEATURES, TARGET], OUTLIER_THRESHOLD).rows
      outer = split_rows(kept, test_size, seed)
      inner = Chapter02.split_train_test(outer.x_train, outer.t_train, test_size: validation_size, seed:)

      scale(inner, outer, kept: kept.size, removed: table.rows.size - kept.size)
    end

    # 訓練データで標準化の平均と標準偏差を求め、訓練・検証・テストの 3 つを同じ値で変換する。
    # inner は訓練データと検証データ、outer はテストデータを持つ第 2 章の分割。
    def scale(inner, outer, kept:, removed:)
      scaler = PolynomialScaler.fit(inner.x_train)
      x_train, x_valid, x_test = [inner.x_train, inner.x_test, outer.x_test].map { |x| scaler.transform(x) }

      BostonSplit.new(x_train:, t_train: inner.t_train, x_valid:, t_valid: inner.t_test, x_test:, t_test: outer.t_test,
                      feature_names: scaler.feature_names, kept:, removed:)
    end

    # 行を特徴量と正解にしてから、第 2 章の関数で訓練用とテスト用に分ける。
    def split_rows(rows, test_size, seed)
      x = rows.map { |row| Chapter02::Features.new(columns: FEATURES, values: FEATURES.map { |c| required_number(row, c) }) }

      Chapter02.split_train_test(x, rows.map { |row| required_number(row, TARGET) }, test_size:, seed:)
    end
  end
end
